In [2]:
import requests
import os

url = "https://api.euron.one/api/v1/euri/embeddings"

headers = {
    "Content-Type": "application/json",
    "Authorization": f"Bearer {os.environ.get('EURI_API_KEY', 'your-api-key-here')}"
}

payload = {
    "input": "The food was delicious and the service was excellent.",
    "model": "text-embedding-3-small"
}

response = requests.post(
    url,
    json=payload,
    headers=headers
)

print(response.json())

{'object': 'list', 'data': [{'object': 'embedding', 'embedding': [-0.03240966796875, -0.00839996337890625, -0.06353759765625, -0.04266357421875, 0.04705810546875, -0.0038661956787109375, -0.0037689208984375, 0.046875, -0.004856109619140625, -0.0386962890625, 0.025665283203125, -0.0294036865234375, 0.005611419677734375, -0.04901123046875, 0.00046515464782714844, 0.00469970703125, 0.0196380615234375, -0.032257080078125, 0.00653839111328125, 0.029937744140625, 0.039642333984375, -0.030609130859375, 0.0066680908203125, 0.0244293212890625, -0.02001953125, 0.02886962890625, -0.038421630859375, 0.02117919921875, -0.0263519287109375, 0.0024394989013671875, 0.033599853515625, 0.0008401870727539062, 0.0408935546875, 0.01123809814453125, -0.0107879638671875, -0.032440185546875, 0.0447998046875, 0.005863189697265625, 0.0306549072265625, 0.0206146240234375, -0.0174102783203125, 0.06292724609375, 0.058013916015625, 0.04266357421875, 0.0211944580078125, -0.000797271728515625, -0.03118896484375, 0.036

In [3]:
result = response.json()
print(result)

{'object': 'list', 'data': [{'object': 'embedding', 'embedding': [-0.03240966796875, -0.00839996337890625, -0.06353759765625, -0.04266357421875, 0.04705810546875, -0.0038661956787109375, -0.0037689208984375, 0.046875, -0.004856109619140625, -0.0386962890625, 0.025665283203125, -0.0294036865234375, 0.005611419677734375, -0.04901123046875, 0.00046515464782714844, 0.00469970703125, 0.0196380615234375, -0.032257080078125, 0.00653839111328125, 0.029937744140625, 0.039642333984375, -0.030609130859375, 0.0066680908203125, 0.0244293212890625, -0.02001953125, 0.02886962890625, -0.038421630859375, 0.02117919921875, -0.0263519287109375, 0.0024394989013671875, 0.033599853515625, 0.0008401870727539062, 0.0408935546875, 0.01123809814453125, -0.0107879638671875, -0.032440185546875, 0.0447998046875, 0.005863189697265625, 0.0306549072265625, 0.0206146240234375, -0.0174102783203125, 0.06292724609375, 0.058013916015625, 0.04266357421875, 0.0211944580078125, -0.000797271728515625, -0.03118896484375, 0.036

In [4]:
result['data'][0]['embedding']

[-0.03240966796875,
 -0.00839996337890625,
 -0.06353759765625,
 -0.04266357421875,
 0.04705810546875,
 -0.0038661956787109375,
 -0.0037689208984375,
 0.046875,
 -0.004856109619140625,
 -0.0386962890625,
 0.025665283203125,
 -0.0294036865234375,
 0.005611419677734375,
 -0.04901123046875,
 0.00046515464782714844,
 0.00469970703125,
 0.0196380615234375,
 -0.032257080078125,
 0.00653839111328125,
 0.029937744140625,
 0.039642333984375,
 -0.030609130859375,
 0.0066680908203125,
 0.0244293212890625,
 -0.02001953125,
 0.02886962890625,
 -0.038421630859375,
 0.02117919921875,
 -0.0263519287109375,
 0.0024394989013671875,
 0.033599853515625,
 0.0008401870727539062,
 0.0408935546875,
 0.01123809814453125,
 -0.0107879638671875,
 -0.032440185546875,
 0.0447998046875,
 0.005863189697265625,
 0.0306549072265625,
 0.0206146240234375,
 -0.0174102783203125,
 0.06292724609375,
 0.058013916015625,
 0.04266357421875,
 0.0211944580078125,
 -0.000797271728515625,
 -0.03118896484375,
 0.036376953125,
 -0.006

In [5]:
len(result['data'][0]['embedding'])

1536

- This length represents embedding vector contains 1536 numbers. This 1536 represents the number of dimensions in the vector.
- result['data'][0]['embedding'] returns the embedding vector generated by the model. 
- len() returns the number of elements in that vector. The output 1536 means the embedding model represents the text as a point in a 1536-dimensional vector space, where each dimension is a learned numerical feature used to capture the semantic meaning of the text.

#### Understanding the Terms:

##### 1. Vector:
- ordered collection of numbers. 
```
Examples -> [10, 20] 
            [5, 8, 12]
            [-0.0324, -0.0084, 0.1234, ...]
```
- Suppose I want to describe a person through the following attributes -> Height, Weight, Age

    I can represent that person as a vector like this, for example : [180, 75, 30]
- In AI, we represent many things as vectors:
    - Images
    - Audio
    - Text
    - Videos

##### 2. Embedding:
- process of converting something (text, image, audio, etc.) into numbers.
- Example:
```
Text -> I love dogs
Embedding model converts it into -> [0.23, -0.44, 0.11, ...]
```
- Why do we need embeddings?
    - Because Computers don't understand "I love dogs", they only understand -> [0.23, -0.44, 0.11, ...]
    - Therefore, 
            ```
                Text -> Embedding Model -> Numbers
            ```

##### 3. Embedding Vector:
- output of an embedding model is called an embedding vector.
- Example: [0.23, -0.44, 0.11, ...]

##### 4. Dimension:
- one position in a vector
- Example: [10, 20]
    - Dimension 1 = 10
    - Dimension 2 = 20

##### 5. Vector Space:
- universe where vectors live.
- Example: 
```
2-D Space for [3,5] means x = 3 and y = 5

Embedding space for [1536 numbers] means -> 
dimension 1
dimension 2
dimension 3
...
dimension 1536
```

##### 6. 1536-Dimensional Vector Space:
- Suppose:
```
embedding = [
  0.12,
  -0.33,
  0.56,
  ...
]
```
contains 1536 numbers. Then len(embedding) returns 1536 which means the vector has 1536 dimensions.

##### 7. Learned Numerical Feature:
- When the embedding model was trained, it read billions of sentences. During training, it learned patterns such as:
    - positivity
    - negativity
    - food-relatedness
    - animal-relatedness
    - sports-relatedness
    - programming-relatedness
and thousands of other hidden patterns.

- Each dimension begins capturing some learned characteristic. For example: [0.45, -0.21, 0.91, ...]
- Imagine:
```
    Dimension 1 → positivity
    Dimension 2 → food-relatedness
    Dimension 3 → animal-relatedness
```
In reality, we don't know exactly what each dimension means. The model learned them automatically. Therefore they are called: Learned Numerical Features because the model learned them during training.

##### 8. Semantic Meaning of the Text:
- Semantic -> means 'Meaning'
- When we talk about the semantic meaning of a sentence, we are asking: "What is this sentence trying to say?" not "What exact words were used?"
- "Semantic Meaning of the Text" means what is the text trying to say.



In [6]:
a = result['data'][0]['embedding']

In [7]:
import requests
import os

url = "https://api.euron.one/api/v1/euri/embeddings"

headers = {
    "Content-Type": "application/json",
    "Authorization": f"Bearer {os.environ.get('EURI_API_KEY', 'your-api-key-here')}"
}

payload = {
    "input": "Rakesh is a software engineer with 5 years of experience in web development.",
    "model": "text-embedding-3-small"
}

response = requests.post(
    url,
    json=payload,
    headers=headers
)

b = response.json()['data'][0]['embedding']

In [8]:
import requests
import os

url = "https://api.euron.one/api/v1/euri/embeddings"

headers = {
    "Content-Type": "application/json",
    "Authorization": f"Bearer {os.environ.get('EURI_API_KEY', 'your-api-key-here')}"
}

payload = {
    "input": "GenAI is transforming the way we interact with technology.",
    "model": "text-embedding-3-small"
}

response = requests.post(
    url,
    json=payload,
    headers=headers
)

c = response.json()['data'][0]['embedding']

In [9]:
import requests
import os

url = "https://api.euron.one/api/v1/euri/embeddings"

headers = {
    "Content-Type": "application/json",
    "Authorization": f"Bearer {os.environ.get('EURI_API_KEY', 'your-api-key-here')}"
}

payload = {
    "input": "What do you do ?",
    "model": "text-embedding-3-small"
}

response = requests.post(
    url,
    json=payload,
    headers=headers
)

inp = response.json()['data'][0]['embedding']

In [10]:
import numpy as np

In [11]:
type(a)

list

In [12]:
a = np.array(a)
b= np.array(b)
c = np.array(c)
inp = np.array(inp)

In [13]:
a,b,c,inp

(array([-0.03240967, -0.00839996, -0.0635376 , ..., -0.00226402,
         0.05899048, -0.03155518], shape=(1536,)),
 array([-0.01965332, -0.00120735, -0.01034546, ..., -0.01198578,
         0.03317261, -0.040802  ], shape=(1536,)),
 array([-0.01169586, -0.00766373,  0.02960205, ..., -0.00185013,
         0.00320625,  0.03092957], shape=(1536,)),
 array([-0.01724243, -0.02185059,  0.0030098 , ...,  0.00585175,
        -0.04165649, -0.00998688], shape=(1536,)))

In [14]:
# linalg.norm() calculates the Euclidean distance between two vectors. The smaller the distance, the more similar the vectors are.
# what does inp-a or inp-b or inp-c mean ? It calculates the difference between the two vectors namely inp and a, inp and b, inp and c respectively.
# The result of inp-a is a new vector that represents the difference between the two vectors. The Euclidean distance is then calculated from this 
# difference vector, which gives a measure of how far apart the two original vectors are in the vector space. A smaller distance indicates that 
# the vectors are more similar, while a larger distance indicates that they are more dissimilar.
dist_inp_a = np.linalg.norm(inp-a)
dist_inp_b = np.linalg.norm(inp-b)
dist_inp_c = np.linalg.norm(inp-c)


In [15]:
dist_inp_a , dist_inp_b, dist_inp_c

# According to the below output, we can conclude that the input "What do you do ?" is more similar to the input 
# "Rakesh is a software engineer with 5 years of experience in web development." because the distance between their embeddings (dist_inp_b) is 
# smaller than the distances to the other two inputs (dist_inp_a and dist_inp_c)

(np.float64(1.3571475919346847),
 np.float64(1.281624224252739),
 np.float64(1.3084758761642523))